dimod/include/dimod/quadratic_model.h

```diff
@@ -583,14 +374,22 @@ class QuadraticModelBase {
     /// Return the number of variables in the quadratic model.
     size_type num_variables() const { return linear_biases_.size(); }
 
-    /// Return the number of interactions in the quadratic model.
+    /**
+     *  Return the number of interactions in the quadratic model.
+     *
+     * `O(num_variables*log(num_variables))` complexity.
+     */
     size_type num_interactions() const {
         size_type count = 0;
+        for (size_type v = 0; v < this->num_variables(); ++v) {
+            count += this->adj_[v].size();
 
-        for (auto it = adj_.begin(); it != adj_.end(); ++it) {
-            count += (*it).size();
+            // account for self-loops
+            auto lb = this->adj_[v].lower_bound(v);
```

Comment:
I think you need to subtract 1 if self loop is present, other wise keep it unchanged.

Reply:
We divide by 2 at the end. So we need to double count the self-loops. In this case we want to count self-loops as interactions. E.g. `i**2` when `i` is an integer is considered an interaction.

Reply:
You can add this reply as a comment :)